File: packages/typescript/src/security/policies.ts
import type { ChatMessage } from "../types";
import type { ContentPolicy } from "./types";

export interface PolicyViolation {
  policyId: string;
  policyName: string;
  action: "block" | "warn";
  matchedPattern?: string;
}

interface CheckContext {
  userId?: string;
  teamId?: string;
  tenantId?: string;
}

export class PolicyEngine {
  private policies: ContentPolicy[];

  constructor(policies: ContentPolicy[]) {
    this.policies = policies;
  }

  /** Get all policies */
  getPolicies(): ContentPolicy[] {
    return [...this.policies]; // return copy to prevent mutation
  }

  /** Add a policy at runtime */
  addPolicy(policy: ContentPolicy): void {
    // Validate policy
    if (!policy.id || !policy.name || !policy.type || !policy.action) {
      throw new Error("Policy must have id, name, type, and action");
    }
    if (this.policies.some(p => p.id === policy.id)) {
      throw new Error(`Policy with id "${policy.id}" already exists`);
    }
    this.policies.push(policy);
  }

  /** Remove a policy by ID */
  removePolicy(id: string): void {
    this.policies = this.policies.filter(p => p.id !== id);
  }

  /** Check messages against all active policies */
  check(messages: ChatMessage[], context: CheckContext): PolicyViolation[] {
    const violations: PolicyViolation[] = [];
    const text = messages.map(m => m.content).join(" ");

    for (const policy of this.policies) {
      // Tenant scope check
      if (policy.tenantId && policy.tenantId !== context.tenantId) continue;

      // Scope check — if policy targets specific teams/users, skip if not in scope
      if (policy.applyTo) {
        const hasUserScope = policy.applyTo.users && policy.applyTo.users.length > 0;
        const hasTeamScope = policy.applyTo.teams && policy.applyTo.teams.length > 0;

        // If teams are specified, user must be in one of those teams
        if (hasTeamScope && (!context.teamId || !policy.applyTo.teams!.includes(context.teamId))) continue;
        // If users are specified, user must be in the list
        if (hasUserScope && (!context.userId || !policy.applyTo.users!.includes(context.userId))) continue;
      }

      let violated = false;
      let matchedPattern: string | undefined;

      switch (policy.type) {
        case "keyword_block":
          if (policy.patterns) {
            for (const kw of policy.patterns) {
              if (text.toLowerCase().includes(kw.toLowerCase())) {
                violated = true;
                matchedPattern = kw;
                break;
              }
            }
          }
          break;

        case "regex_block":
          if (policy.regex) {
            // ReDoS protection — reject patterns with nested quantifiers
            if (/\([^)]*[+*][^)]*\)[+*]/.test(policy.regex)) break;
            if (/(\.\*){3,}/.test(policy.regex)) break;
            try {
              const regex = new RegExp(policy.regex, "gi");
              const match = regex.exec(text);
              if (match) {
                violated = true;
                matchedPattern = match[0];
              }
            } catch { /* invalid regex — skip */ }
          }
          break;

        case "topic_restriction":
          if (policy.patterns) {
            const lowerText = text.toLowerCase();
            for (const topic of policy.patterns) {
              if (lowerText.includes(topic.toLowerCase())) {
                violated = true;
                matchedPattern = topic;
                break;
              }
            }
          }
          break;

        case "max_tokens":
          if (policy.maxTokens) {
            // Conservative estimate: 1 token ≈ 3 chars (errs on the side of blocking)
            // Actual tokenization varies: English prose ~4 chars/token, code/short words ~2-3
            const estimatedTokens = Math.ceil(text.length / 3);
            if (estimatedTokens > policy.maxTokens) {
              violated = true;
              matchedPattern = `${estimatedTokens} tokens (max: ${policy.maxTokens})`;
            }
          }
          break;
      }

      if (violated) {
        violations.push({
          policyId: policy.id,
          policyName: policy.name,
          action: policy.action,
          matchedPattern,
        });
      }
    }

    return violations;
  }
}

File: packages/python/bulwark_ai/security/policies.py
"""
Content Policy Engine.

Evaluates chat messages against keyword, regex, topic, and token-limit policies.
Supports tenant scoping and per-user/team targeting.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from typing import Literal

from ..types import ChatMessage, ContentPolicy


# ReDoS detection patterns (same as PII module)
_REDOS_NESTED_QUANTIFIER = re.compile(r"\([^)]*[+*][^)]*\)[+*]")
_REDOS_EXCESSIVE_WILDCARDS = re.compile(r"(\.\*){3,}")


@dataclass
class PolicyViolation:
    policy_id: str
    policy_name: str
    action: Literal["block", "warn"]
    matched_pattern: str | None = None


class PolicyEngine:
    """
    Evaluates messages against content policies.

    Supports four policy types:
    - keyword_block: block/warn on keyword matches
    - regex_block: block/warn on regex matches (with ReDoS protection)
    - topic_restriction: block/warn on topic keyword matches
    - max_tokens: block/warn when estimated token count exceeds limit
    """

    def __init__(self, policies: list[ContentPolicy] | None = None) -> None:
        self._policies: list[ContentPolicy] = list(policies or [])

    def get_policies(self) -> list[ContentPolicy]:
        """Return a copy of all policies."""
        return list(self._policies)

    def add_policy(self, policy: ContentPolicy) -> None:
        """Add a policy at runtime."""
        if not policy.id or not policy.name or not policy.type or not policy.action:
            raise ValueError("Policy must have id, name, type, and action")
        if any(p.id == policy.id for p in self._policies):
            raise ValueError(f'Policy with id "{policy.id}" already exists')
        self._policies.append(policy)

    def remove_policy(self, policy_id: str) -> None:
        """Remove a policy by ID."""
        self._policies = [p for p in self._policies if p.id != policy_id]

    def check(
        self,
        messages: list[ChatMessage],
        *,
        user_id: str | None = None,
        team_id: str | None = None,
        tenant_id: str | None = None,
    ) -> list[PolicyViolation]:
        """Check messages against all active policies."""
        violations: list[PolicyViolation] = []
        text = " ".join(m.content for m in messages)

        for policy in self._policies:
            # Tenant scope check
            if policy.tenant_id and policy.tenant_id != tenant_id:
                continue

            # Scope check -- if policy targets specific teams/users, skip if not in scope
            if policy.apply_to:
                users = policy.apply_to.get("users", [])
                teams = policy.apply_to.get("teams", [])
                if teams and (not team_id or team_id not in teams):
                    continue
                if users and (not user_id or user_id not in users):
                    continue

            violated = False
            matched_pattern: str | None = None

            if policy.type == "keyword_block":
                if policy.patterns:
                    lower_text = text.lower()
                    for kw in policy.patterns:
                        if kw.lower() in lower_text:
                            violated = True
                            matched_pattern = kw
                            break

            elif policy.type == "regex_block":
                if policy.regex:
                    # ReDoS protection
                    if _REDOS_NESTED_QUANTIFIER.search(policy.regex):
                        continue
                    if _REDOS_EXCESSIVE_WILDCARDS.search(policy.regex):
                        continue
                    try:
                        regex = re.compile(policy.regex, re.IGNORECASE)
                        m = regex.search(text)
                        if m:
                            violated = True
                            matched_pattern = m.group(0)
                    except re.error:
                        pass  # Invalid regex -- skip

            elif policy.type == "topic_restriction":
                if policy.patterns:
                    lower_text = text.lower()
                    for topic in policy.patterns:
                        if topic.lower() in lower_text:
                            violated = True
                            matched_pattern = topic
                            break

            elif policy.type == "max_tokens":
                if policy.max_tokens:
                    # Conservative estimate: 1 token ~ 3 chars
                    estimated_tokens = math.ceil(len(text) / 3)
                    if estimated_tokens > policy.max_tokens:
                        violated = True
                        matched_pattern = f"{estimated_tokens} tokens (max: {policy.max_tokens})"

            if violated:
                violations.append(PolicyViolation(
                    policy_id=policy.id,
                    policy_name=policy.name,
                    action=policy.action,
                    matched_pattern=matched_pattern,
                ))

        return violations
